finalitygadget/finalitygadget.go

```diff
@@ -385,81 +389,83 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
-	if localBlock != nil && localBlock.BlockHeight >= block.BlockHeight {
-		block = &types.Block{
-			BlockHeight:    localBlock.BlockHeight,
-			BlockHash:      localBlock.BlockHash,
-			BlockTimestamp: localBlock.BlockTimestamp,
-		}
-	} else {
-		// throw if block height is 0
-		if block.BlockHeight == 0 {
-			return fmt.Errorf("block height 0")
-		}
-
-		// Check the block is finalized using sdk client
-		isFinal, err := fg.QueryIsBlockBabylonFinalized(block)
-		// If not finalized, throw error
-		if !isFinal {
-			return fmt.Errorf("block %d should be finalized according to client but is not", block.BlockHeight)
-		}
-		if err != nil {
-			return fmt.Errorf("error checking block %d: %v", block.BlockHeight, err)
-		}
-		// If finalised, store the block in DB and set the last finalized block
-		err = fg.InsertBlock(block)
-		if err != nil {
-			return fmt.Errorf("error storing block %d: %v", block.BlockHeight, err)
+		if errors.Is(err, types.ErrBtcStakingNotActivated) {
+			fg.logger.Info("BTC staking not yet activated, waiting...")
+			return nil
 		}
+		return fmt.Errorf("error querying BTC staking activation timestamp: %w", err)
 	}
 
-	// Start service at block height
-	fg.logger.Info("Starting finality gadget at block", zap.Uint64("block_height", block.BlockHeight))
-
-	// Set the curr finalized block in memory
-	fg.currHeight = block.BlockHeight
+	for height := fg.lastProcessedHeight + 1; height <= latestFinalizedHeight; height++ {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			block, err := fg.queryBlockByHeight(int64(height))
+			if err != nil {
+				return fmt.Errorf("error getting block at height %d: %w", height, err)
+			}
 
-	return nil
-}
+			// only process blocks after the btc staking is activated
+			if block.BlockTimestamp < btcStakingActivatedTimestamp {
+				fg.logger.Info("Skipping block before BTC staking activation", zap.Uint64("block_height", block.BlockHeight))
+				fg.lastProcessedHeight = block.BlockHeight
+				continue
+			}
 
-func (fg *FinalityGadget) handleBlock(block *types.Block) {
-	// while block is not finalized, recheck if block is finalized every `retryInterval` seconds
-	// if finalized, store the block in DB and set the last finalized block
-	for {
-		// Check if block is finalized
-		isFinal, err := fg.QueryIsBlockBabylonFinalized(block)
-		if err != nil {
-			fg.logger.Fatal("Error checking block", zap.Uint64("block_height", block.BlockHeight), zap.Error(err))
-			return
-		}
-		if isFinal {
-			err = fg.InsertBlock(block)
+			// Query local DB for last block processed
+			localBlock, err := fg.db.QueryLatestFinalizedBlock()
 			if err != nil {
-				fg.logger.Fatal("Error storing block", zap.Uint64("block_height", block.BlockHeight), zap.Error(err))
+				return fmt.Errorf("error getting latest block from db: %v", err)
 			}
-			return
-		}
 
-		// Sleep for `PollInterval` seconds
-		time.Sleep(fg.pollInterval * time.Second)
+			// store the first babylon finalized block to local DB
```

Comment:
suggest clarifying this comment "store the first babylon finalized block on startup, then store any new blocks to local DB"

Reply:
actually do we need the `localBlock.BlockHeight < block.BlockHeight` condition in L426? arent all blocks in this loop going to be gt the local tip since it starts from `lastProcessedHeight + 1`

Reply:
yes, it is not needed b/c we set the `lastProcessedHeight` with the local tip when starting finality gadget